### PR TITLE
Added method to create codec for an Enumeration

### PR DESCRIPTION
**Pull request by ajaychandran**

### CODE REVIEW

**Comment by ajaychandran** [comment]
Hi Michael,

I am still quite new to the GitHub process and not sure if I am following the correct procedure.

Please let me know if you need me to resubmit anything. For now, I will continue to commit to my branch.

Regards,
Ajay

File: src/main/scala/scodec/codecs/package.scala [hunk 1]
```diff
@@ -1101,33 +1101,43 @@ package object codecs {
...
    * @return
    */
-  def checksummed[A](target: Codec[A], checksum: Encoder[BitVector], rangeSize: Decoder[Long], rangePadding: Long): Codec[A] =
-    new ChecksumCodec(target, checksum, rangeSize, rangePadding)
+  def filtered[A](filter: Codec[BitVector], codec: Codec[A]): Codec[A] = new Codec[A] {
+      def encode(value: A): Attempt[BitVector] = codec.encode(value) flatMap filter.encode
+      def sizeBound: SizeBound = filter.sizeBound
+      def decode(bits: BitVector): Attempt[DecodeResult[A]] =
+        filter.decode(bits)
+          .fold(e => Attempt.failure(e), r => codec.decode(r.value)
```
Comment (mpilquist) [inline src/main/scala/scodec/codecs/package.scala]: This first `fold` can be replaced with a `flatMap` and the second `fold` can be replaced with a `map`.

**Comment by mpilquist** [comment]
Thanks! Looks good -- just one minor comment about fold vs flatMap.

File: src/test/scala/scodec/codecs/ChecksumCodecTest.scala [hunk 2]
```diff
@@ -1,17 +1,21 @@
 package scodec.codecs
 
+import scodec._
 import scodec.bits._
-import scodec.{Attempt, CodecSuite}
 
+/**
```
Comment (mpilquist) [inline src/test/scala/scodec/codecs/ChecksumCodecTest.scala]: Silly style thing, but could you remove this comment?

File: src/main/scala/scodec/codecs/package.scala (hunk 1 above)
Comment (ajaychandran) [inline src/main/scala/scodec/codecs/package.scala]: Okay, but I think the second fold should also be replaced with `flatMap` in order to preserve the remainder (if any) from `codec`.

```
      def decode(bits: BitVector): Attempt[DecodeResult[A]] =
        filter.decode(bits)
          .flatMap(r => codec.decode(r.value)
            .flatMap(a => Attempt.successful(a.mapRemainder(_ ++ r.remainder))))
```

Comment (mpilquist) [inline src/main/scala/scodec/codecs/package.scala]: `.flatMap(a => Attempt.successful(f(a))` is equivalent to `.map(a => f(a))` -- so in this case, the last `flatMap` could be `map(_.mapRemainder(...))`

**Comment by mpilquist** [comment]
@ajaychandran Thanks for the PR! I merged this one but then found a way to further generalize `checksummed` by separating the framing logic from the checksumming.

Take a look at the change in usage: https://github.com/scodec/scodec/blob/5e0ef451b7eebb999fef210c1833facedc43a136/src/test/scala/scodec/codecs/ChecksumCodecTest.scala

The full change is here: https://github.com/scodec/scodec/commit/5e0ef451b7eebb999fef210c1833facedc43a136

Regarding the delimited codec, could you open a new PR for that one? I'll add comments inline on the PR rather than on the gist and via email.

Cheers

**Pull request merged**